src/devinterp/slt/llc.py

```diff
@@ -0,0 +1,186 @@
+import inspect
+import warnings
+from typing import Callable, Dict, List, Literal, Optional, Type, Union
+
+import numpy as np
+import pandas as pd
+import torch
+from torch.utils.data import DataLoader
+
+from devinterp.optim.sgld import SGLD
+from devinterp.slt.estimators import Estimator
+from devinterp.slt.sampler import sample
+
+
+class LLCEstimator(Estimator):
+    def __init__(self, num_chains: int, num_draws: int, n: int, device="cpu"):
+        self.num_chains = num_chains
+        self.num_draws = num_draws
+        self.losses = np.zeros((num_chains, num_draws), dtype=np.float32)
+
+        self.n = torch.tensor(n, dtype=torch.float32).to(device)
+        self.llc_per_chain = torch.zeros(num_chains, dtype=torch.float32).to(device)
+        self.llc_mean = torch.tensor(0., dtype=torch.float32).to(device)
+        self.llc_std = torch.tensor(0., dtype=torch.float32).to(device)
+
+        self.device = device
+
+    def update(self, chain: int, draw: int, loss: float):
+        self.losses[chain, draw] = loss 
+
+    @property
+    def init_loss(self):
+        return self.losses[0, 0]
+
+    def finalize(self):
+        avg_losses = self.losses.mean(axis=1)
+        self.llc_per_chain = (self.n / self.n.log()).detach().cpu().numpy() * (avg_losses - self.init_loss)
+        self.llc_mean = self.llc_per_chain.mean()
+        self.llc_std = self.llc_per_chain.std()
+        
+    def sample(self):
+        return {
+            "llc/mean": self.llc_mean.item(),
+            "llc/std": self.llc_std.item(),
+            **{f"llc-chain/{i}": self.llc_per_chain[i].item() for i in range(self.num_chains)},
+            "loss/trace": self.losses,
+        }
+    
+    def __call__(self, chain: int, draw: int, loss: float):
+        self.update(chain, draw, loss)
+
+
+class OnlineLLCEstimator(Estimator):
+    def __init__(self, num_chains: int, num_draws: int, n: int, device="cpu"):
+        self.num_chains = num_chains
+        self.num_draws = num_draws
+
+        self.losses = np.zeros((num_chains, num_draws), dtype=torch.float32)
```

Comment:
`np.zeros` -> `torch.zeros` or it throws an error bc of `dtype=torch.float32`

I think this is also fine if you do `np.zeros((num_chains, num_draws))`, it seems like if you try to insert a tensor then python will just cast it to an `np.float64`